File: src/1290.convert_binary_number_in_a_linked_list_to_integer.kt
import other.ListNode

fun main() {
    println(getDecimalValue2(ListNode(1).apply {
        this.next = ListNode(0).apply {
            this.next = ListNode(1)
        }
    }))
    println(getDecimalValue2(ListNode(0)))
}

fun getDecimalValue(head: ListNode?): Int {
    val builder = StringBuilder()

    var node = head;
    while (node != null) {
        builder.append(node.`val`)
        node = node.next
    }

    return builder.toString().toBigInteger(2).intValueExact()
}

fun getDecimalValue2(head: ListNode?): Int {
    var count = 0

    var node = head
    while (node != null) {
        count = count.shl(1)
        if (node.`val` > 0) {
            count++;
        }

        node = node.next
    }

    return count
}

File: lib/1290.convert_binary_number_in_a_linked_list_to_integer.dart
import 'other/list_node.dart';

void main() {
  print(getDecimalValue2(ListNode(1)..next = (ListNode(0)..next = ListNode(1))));
  print(getDecimalValue2(ListNode(0)));
}

int getDecimalValue(ListNode? head) {
  var str = "";

  while (head != null) {
    str += "${head.val}";
    head = head.next;
  }

  return int.parse(str, radix: 2);
}

int getDecimalValue2(ListNode? head) {
  var count = 0;

  while (head != null) {
    count = count << 1;

    if (head.val > 0) {
      count++;
    }
    head = head.next;
  }

  return count;
}
